File: library/src/main/java/wannabe/grid/iterators/InBoundsIterator.java
package wannabe.grid.iterators;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import wannabe.Bounds;
import wannabe.Voxel;
import wannabe.grid.Grid;

public final class InBoundsIterator implements Iterator<Voxel> {
  private final Iterator<Voxel> realIterator;

  public InBoundsIterator(Grid source, Bounds bounds) {
    List<Voxel> inBounds = new ArrayList<>(source.size());
    for (Voxel voxel : source) {
      if (bounds.contains(voxel.position)) {
        inBounds.add(voxel);
      }
    }
    realIterator = inBounds.iterator();
  }

  @Override public boolean hasNext() {
    return realIterator.hasNext();
  }

  @Override public Voxel next() {
    return realIterator.next();
  }

  @Override public void remove() {
    // TODO should I support remove?
    realIterator.remove();
  }
}

File: library/src/main/java/wannabe/grid/iterators/InBoundsIterator.kt
package wannabe.grid.iterators

import wannabe.Bounds
import wannabe.Voxel
import wannabe.grid.Grid
import java.util.ArrayList

// TODO this should be a streaming solution not a caching one
class InBoundsIterator(
  source: Grid,
  bounds: Bounds
) : MutableIterator<Voxel> {
  private val realIterator: MutableIterator<Voxel>

  init {
    val inBounds = source.filter { bounds.contains(it.position) }
        .toMutableList()
    realIterator = inBounds.iterator()
  }

  override fun hasNext(): Boolean = realIterator.hasNext()

  override fun next(): Voxel = realIterator.next()

  // TODO should I support remove?
  override fun remove() = realIterator.remove()
}
